Fetch obs from P1D and PT10M

guillaumevernieres opened the pull request:
### Description
All of the marine SST retrieval obs are stored in 10 mn ioda files, and require to be concatenated post r2d2 fetch. This PR adds the tools to do so.

- fixes #182 

**Possible breaking changes:** `stage.obs` seems to be only used by the marine application so presumably it shouldn't be an issue, but the scope of what `stage.obs` does has been changed. If you are using that function but your use of it isn't covered in the ctests, then this PR will break your stuff.

#### Other things done not really related to the issue above:
- Changed scales for **C** in **B**
- Fixed bug in 3dvar `yaml`

### Danger!
- ~This PR, which should probably have been a draft, will probably break the stand alone applications. The environment variable ```JEDI_LIB``` pointing to ```<path to>/build/lib``` needs to be set. This is taken care of in the ctests, but not in the applications (yet, to do before merging ...).~ 
- ```stage.py``` has a dependency on ioda ... 
- Some would say the ```import``` of  ```ioda_conv_engines``` [here](https://github.com/NOAA-EMC/GDASApp/blob/52a5157fb9cc27f9c207d7bae5e770aff283bcc3/ush/ufsda/stage.py#L19) is ~clunky!~ great!

guillaumevernieres committed:
wip

guillaumevernieres committed:
reverted some of Jiarui's changes

guillaumevernieres committed:
concatenate ioda pt10m with xarray

guillaumevernieres committed:
Merge branch 'develop' into feature/pt10m

guillaumevernieres committed:
writing with ioda

guillaumevernieres committed:
works but needs tidy

guillaumevernieres committed:
Merge branch 'develop' into feature/pt10m

guillaumevernieres committed:
minor cleanup

guillaumevernieres committed:
removed hard-coded path

guillaumevernieres committed:
pip install of xarray

guillaumevernieres committed:
wip

guillaumevernieres committed:
fixed boolean dyslexia?

guillaumevernieres committed:
0%@&

guillaumevernieres committed:
...

guillaumevernieres committed:
Merge branch 'develop' into feature/pt10m

guillaumevernieres committed:
missing lib path

CoryMartin-NOAA commented on test/CMakeLists.txt:
```diff
@@ -93,6 +93,12 @@ install(FILES ${test_reference}
 
 ##### unit tests
+# some of the tests below depend on pyioda and need to be skipped if BUILD_GDASBUNDLE=OFF
+set(SKIPTEST TRUE)
```
why not just wrap these tests in a `if (${BUILD_GDASBUNDLE})`?

guillaumevernieres replied:
good question @CoryMartin-NOAA , still can't remember why I did that ... Will do.

CoryMartin-NOAA commented on ush/ufsda/stage.py:
```diff
@@ -1,19 +1,94 @@
+# TODO: There's probably a better way to do this
+jedi_lib = os.getenv('JEDI_LIB')
+sys.path.append(os.path.join(jedi_lib, 'pyiodaconv'))
+sys.path.append(glob.glob(os.path.join(jedi_lib, 'python3.*/pyioda'))[0])
```
All of this is fine for temporary, but you should be aware, that `stage` is likely going to become generic and not R2D2y soon... I think we are probably going to have to talk about R2D2's role here going forward

guillaumevernieres replied:
@CoryMartin-NOAA , I know we said I should just add the paths of `pyioda` and `pyiodaconv` to `PYTHONPATH`, but I can't figure out how to do this for the ctests using the `cmake_generator_expression` `TARGET_FILE_DIR`, which is what's used in `iodaconv` ... The issue is that the path is python version dependent, so the path to pyioda isn't resolved at config time ... Asking for a pass on that issue ... Pretty please?

CoryMartin-NOAA replied:
I see... well that's a problem. Can you use the cmake variable `Python_VERSION` combined with what is here:
https://github.com/NOAA-EMC/GDASApp/blob/38ca1e9267507cea9ec42f2b50045a2706a0876d/test/atm/CMakeLists.txt#L25

If not, then I will give you a pass.

CoryMartin-NOAA replied:
maybe you will want `Python_VERSION_MAJOR`

guillaumevernieres replied:
Dam ... yes, that should work. ```Python_VERSION_MINOR``` returns the `9` ...

RussTreadon-NOAA reviewed (dismissed):
These changes appear to be largely limited to GDASApp-soca and, as such, should not interfere with other apps.   I am not sure how we plan on using `stage.py` in GDASApp-atm.  Approve.

guillaumevernieres committed:
path2iodastuff

guillaumevernieres committed:
Merge branch 'develop' into feature/pt10m

guillaumevernieres commented:
> These changes appear to be largely limited to GDASApp-soca and, as such, should not interfere with other apps. I am not sure how we plan on using `stage.py` in GDASApp-atm. Approve.

@RussTreadon-NOAA , it's probably breaking the stand alone apps since I introduced dependencies to `pyioda` and `pyiodaconv`. We need to add the path to these libs to `PYTHONPATH`. I have a simple fix for `soca` but I haven't looked at the `atm` app yet ...

RussTreadon-NOAA commented:
> > These changes appear to be largely limited to GDASApp-soca and, as such, should not interfere with other apps. I am not sure how we plan on using `stage.py` in GDASApp-atm. Approve.
> 
> @RussTreadon-NOAA , it's probably breaking the stand alone apps since I introduced dependencies to `pyioda` and `pyiodaconv`. We need to add the path to these libs to `PYTHONPATH`. I have a simple fix for `soca` but I haven't looked at the `atm` app yet ...

Thank you @guillaumevernieres for the heads up.

guillaumevernieres committed:
path to ioda stuff

guillaumevernieres commented:
> > > These changes appear to be largely limited to GDASApp-soca and, as such, should not interfere with other apps. I am not sure how we plan on using `stage.py` in GDASApp-atm. Approve.
> > 
> > 
> > @RussTreadon-NOAA , it's probably breaking the stand alone apps since I introduced dependencies to `pyioda` and `pyiodaconv`. We need to add the path to these libs to `PYTHONPATH`. I have a simple fix for `soca` but I haven't looked at the `atm` app yet ...
> 
> Thank you @guillaumevernieres for the heads up.

@RussTreadon-NOAA , @CoryMartin-NOAA , I'm trying to figure out a way to not "break" the `run_jedi_exe.py` for the `atm` applications, but I'm afraid we'll have to add stuff to `PYTHONPATH` before running. 

This:
```python
    python_version = 'python3.'+sys.version_info[1]
    pyioda_lib = os.path.join(all_config_dict['GDASApp home'], 'build', 'lib', python_version, 'pyioda')
    pyiodaconv_lib = os.path.join(all_config_dict['GDASApp home'], 'build', 'lib', 'pyiodaconv')
    sys.path.append(pyioda_lib)
    sys.path.append(pyiodaconv_lib)
```
added to `run_jedi_exe.py` does not do the trick. `sys_path.append`'s scope is limited to `run_jedi_exe.py` and `stage.py` cannot find pyioda/pyiodaconv. Let me know if you can think of an other way to do this.

My plan is to generate a `load_envar.sh` script for the `atm` app and source it in `submit_job.sh`. That's what I do for the soca app ... Let me know if that's acceptable.

CoryMartin-NOAA commented:
@guillaumevernieres I think that is fine with me, especially since python virtual environments will need to be loaded anyways.

guillaumevernieres committed:
sourcing envar for atm

guillaumevernieres committed:
cn

RussTreadon-NOAA commented:
I cloned and built a fresh copy of [feature/pt10m](https://github.com/NOAA-EMC/GDASApp/tree/feature/pt10m) at 6a75f8d on Hera.   run_jedi_exe.py fails with the following traceback when attempting to run atm 3dvar.

```
(gdasapp) Hera(hfe05):/scratch1/NCEPDEV/da/Russ.Treadon/git/GDASApp/pr190/ush$ ./run_jedi_exe.py -c examples/run_jedi_exe/3dvar_hera.yaml
2022-11-21 16:44:02:INFO:Loading configuration from examples/run_jedi_exe/3dvar_hera.yaml
2022-11-21 16:44:02:INFO:Created working directory /scratch2/NCEPDEV/stmp1/Russ.Treadon//gdas_single_test_3dvar_pr190
2022-11-21 16:44:03:INFO:Note: detected 80 virtual cores but NumExpr set to maximum of 64, check "NUMEXPR_MAX_THREADS" environment variable.
2022-11-21 16:44:03:INFO:Note: NumExpr detected 80 cores but "NUMEXPR_MAX_THREADS" not set, so enforcing safe limit of 8.
Traceback (most recent call last):
  File "./run_jedi_exe.py", line 229, in <module>
    run_jedi_exe(args.config)
  File "./run_jedi_exe.py", line 62, in run_jedi_exe
    import ufsda
  File "/scratch1/NCEPDEV/da/Russ.Treadon/git/GDASApp/pr190/ush/ufsda/__init__.py", line 3, in <module>
    import ufsda.stage
  File "/scratch1/NCEPDEV/da/Russ.Treadon/git/GDASApp/pr190/ush/ufsda/stage.py", line 18, in <module>
    import ioda_conv_engines as iconv
ModuleNotFoundError: No module named 'ioda_conv_engines'
```

guillaumevernieres commented:
Thanks for testing @RussTreadon-NOAA ! I should have mentioned it isn't ready yet. I'll ping you when my fix actually works.

guillaumevernieres commented:
@RussTreadon-NOAA , @CoryMartin-NOAA , What I have in the last commit is not enough. We'll still need to change the PYTHONPATH before running the app.
```console
(gdasapp) bash-4.2$ export PYTHONPATH=$PYTHONPATH:....GDASApp/build/lib/python3.7/pyioda:......./GDASApp/build/lib/pyiodaconv
```
Not ideal ...

RussTreadon-NOAA commented:
> @RussTreadon-NOAA , @CoryMartin-NOAA , What I have in the last commit is not enough. We'll still need to change the PYTHONPATH before running the app.
> 
> ```
> (gdasapp) bash-4.2$ export PYTHONPATH=$PYTHONPATH:....GDASApp/build/lib/python3.7/pyioda:......./GDASApp/build/lib/pyiodaconv
> ```
> 
> Not ideal ...

I now see a different error

> @RussTreadon-NOAA , @CoryMartin-NOAA , What I have in the last commit is not enough. We'll still need to change the PYTHONPATH before running the app.
> 
> ```
> (gdasapp) bash-4.2$ export PYTHONPATH=$PYTHONPATH:....GDASApp/build/lib/python3.7/pyioda:......./GDASApp/build/lib/pyiodaconv
> ```
> 
> Not ideal ...

Got further but fv3jedi_var.x failed for 3dvar atm with

```
ufo_geovals_get_var geopotential_height doesnt exist
Abort(1) on node 1 (rank 1 in comm 0): application called MPI_Abort(MPI_COMM_WORLD, 1) - process 1
```

Does this point to the need to update something in light of last week's code sprint?

CoryMartin-NOAA commented:
The code sprint changes are not yet in develop. Plus GDASApp should still be checking out a fixed commit of all (most?) JEDI repos.

Very odd that geopotential height does not exist...

CoryMartin-NOAA commented:
@guillaumevernieres would this be fixed by a `make install` that puts the library somewhere that is more static/constant? Or are we always going to deal with the python3.X issue?

guillaumevernieres commented:
> @guillaumevernieres would this be fixed by a `make install` that puts the library somewhere that is more static/constant? Or are we always going to deal with the python3.X issue?

Maybe @CoryMartin-NOAA. The issue now is that we'd need to change the pythonpath in the jjob in the g-w, or somewhere ... I think we need to get back to relative paths for these 2 libs in stage.py, similarly to what's done in the iodaconv scripts.

guillaumevernieres commented:
Thanks for testing @RussTreadon-NOAA looks like a problem unrelated to this PR.

RussTreadon-NOAA commented:
> Thanks for testing @RussTreadon-NOAA looks like a problem unrelated to this PR.

Agreed, I am encountering issues outside the scope of this PR.

If I comment out `- $<< $(OBS_YAML_DIR)/sfc.yaml` in `parm/atm/obs/lists/gdas_prototype.yaml`, then fv3jedi_var.x built from feature/pt10m gets further but fails when processing cris-fsr obs.   The cris issue appears to be memory related.   The failure message is
```
[h20c26:134962:0:134962] Caught signal 7 (Bus error: nonexistent physical address)
[h20c26:134959:0:134959] Caught signal 7 (Bus error: nonexistent physical address)
srun: error: h20c26: task 4: Killed
srun: launch/slurm: _step_signal: Terminating StepId=38046916.0
slurmstepd: error: *** STEP 38046916.0 ON h20c26 CANCELLED AT 2022-11-21T20:53:58 ***
```
Comment out both n20 and npp cris-fsr.   Job runs but cost function is order 1e+80.   Remove ompsnp_npp (layer ozone) and retain /ompstc8_npp (total column ozone) and cost function returns to reasonable value.   The nonlinear Jo for ompsnp_npp is not huge (473.306206) so suspect problem with ompsnp_npp gradient calculation.

CoryMartin-NOAA commented:
I vaguely remember issues with both sfc and ozone in the var from weeks ago? I guess that was never fixed. I hope I didn't promise to fix that and forgot...

guillaumevernieres committed:
reverted iodaconv stuff

guillaumevernieres committed:
Merge branch 'develop' into feature/pt10m

guillaumevernieres committed:
reverted more stand alone app stuff

guillaumevernieres commented:
@CoryMartin-NOAA , @RussTreadon-NOAA , I reverted to finding the path to `iodaconv` stuff in `stage.yaml`. It's based on assuming where the library could be relative to `stage.py` ... Not great, but until we pip install the repo ... or something, I say this is good enough.
@RussTreadon-NOAA , I have not tested the atm app, but it works with soca without specifying the path to the iodaconv lib. Any chance you could retest? (after clearing PYTHONPATH of the 2 libs).

_ps: It should work with the g-w as well but I have not tested yet ... Will try now, or tomorrow._

guillaumevernieres committed:
more dyslexia

RussTreadon-NOAA commented:
Used feature/pt10m @ 4472763 to execute `./run_jedi_exe.py -c examples/run_jedi_exe/3dvar_hera.yaml`.   The job ran to completion with obs types noted above commented out in the obs yaml.

Nothing was added to `PYTHONPATH` in this test.   `run_jedi_exe.py` ran with
```
(gdasapp) Hera(hfe11):/scratch1/NCEPDEV/da/Russ.Treadon/git/GDASApp/pr190/ush$ echo $PYTHONPATH
/scratch1/NCEPDEV/global/spack-stack/spack-stack-v1/envs/skylab-2.0.0-intel-2021.5.0/install/intel/2021.5.0/py-pybind11-2.8.1-jnhqvzf/lib/python3.9/site-packages:/scratch1/NCEPDEV/global/spack-stack/spack-stack-v1/envs/skylab-2.0.0-intel-2021.5.0/install/intel/2021.5.0/bufr-11.7.1-a7u4hj2/lib/python3.9/site-packages
```
in the environment.

guillaumevernieres commented:
> Used feature/pt10m @ [4472763](https://github.com/NOAA-EMC/GDASApp/commit/44727635a8f4da1e0579f548c84e66f40f8bf20a) to execute `./run_jedi_exe.py -c examples/run_jedi_exe/3dvar_hera.yaml`. The job ran to completion with obs types noted above commented out in the obs yaml.
> 
> Nothing was added to `PYTHONPATH` in this test. `run_jedi_exe.py` ran with
> 
> ```
> (gdasapp) Hera(hfe11):/scratch1/NCEPDEV/da/Russ.Treadon/git/GDASApp/pr190/ush$ echo $PYTHONPATH
> /scratch1/NCEPDEV/global/spack-stack/spack-stack-v1/envs/skylab-2.0.0-intel-2021.5.0/install/intel/2021.5.0/py-pybind11-2.8.1-jnhqvzf/lib/python3.9/site-packages:/scratch1/NCEPDEV/global/spack-stack/spack-stack-v1/envs/skylab-2.0.0-intel-2021.5.0/install/intel/2021.5.0/bufr-11.7.1-a7u4hj2/lib/python3.9/site-packages
> ```
> 
> in the environment.

What's that they say? "First try!"
Thanks for testing @RussTreadon-NOAA .

RussTreadon-NOAA reviewed (approved):
Changes in this PR do not break atm 3dvar.   Approve.
